xdsl/irdl/declarative_assembly_format.py

```diff
@@ -438,27 +438,41 @@ def parse(self, parser: Parser, state: ParsingState) -> None:
 
     def print(self, printer: Printer, state: PrintingState, op: IRDLOperation) -> None:
         if self.print_properties:
-            if (
-                not (set(op.attributes.keys()) | set(op.properties.keys()))
-                - self.reserved_attr_names
-            ):
-                return
             if any(name in op.attributes for name in op.properties):
                 raise ValueError(
                     "Cannot print attributes and properties with the same name "
-                    "in a signle dictionary"
+                    "in a single dictionary"
                 )
+            op_def = op.get_irdl_definition()
+            dictionary = op.attributes | op.properties
+            default_names = set(
+                name
+                for name, d in (op_def.properties | op_def.attributes).items()
+                if d.default_value is not None
+                and dictionary.get(name) == d.default_value
+            )
+            reserved_or_default = self.reserved_attr_names | default_names
+            if not (set(dictionary.keys())) - reserved_or_default:
+                return
             printer.print_op_attributes(
-                op.attributes | op.properties,
-                reserved_attr_names=self.reserved_attr_names,
+                dictionary,
+                reserved_attr_names=reserved_or_default,
                 print_keyword=self.with_keyword,
             )
         else:
-            if not set(op.attributes.keys()) - self.reserved_attr_names:
+            op_def = op.get_irdl_definition()
+            default_names = set(
+                name
+                for name, d in op_def.attributes.items()
+                if d.default_value is not None
+                and op.attributes.get(name) == d.default_value
+            )
+            reserved_or_default = self.reserved_attr_names | default_names
+            if not set(op.attributes.keys()) - reserved_or_default:
```

Comment:
nit:
```suggestion
            reserved_or_default = self.reserved_attr_names.update(
                name
                for name, d in op_def.attributes.items()
                if d.default_value is not None
                and op.attributes.get(name) == d.default_value
            )
            if not reserved_or_default.is_superset(op.attributes.keys()):
```
Two proposals, each of which avoids a set construction, seems a tiny bit cleaner

Reply:
update is a mutating function right? (I can't mutate self.reserved_attr_names)

Reply:
My bad, I meant `union`